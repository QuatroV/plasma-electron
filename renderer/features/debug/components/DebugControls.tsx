import { useState } from "react";
import { HiRefresh } from "react-icons/hi";
import { IoIosArrowForward, IoIosExit } from "react-icons/io";
import { MdOutlineClear } from "react-icons/md";
import { TbPlayerStop } from "react-icons/tb";
import {
  VscDebugStart,
  VscDebugStepInto,
  VscDebugStepOver,
} from "react-icons/vsc";

import Button from "../../../components/Button";
import Input from "../../../components/Input";
import Tooltip from "../../../components/Tooltip";
import clsxm from "../../../utils/clsxm";
import useDebug from "../hooks/useDebug";
import useDebugStore from "../stores/debugStore";
import { transformRegistersResponse } from "../utils";

const DebugControls = () => {
  const { runDbgCmd } = useDebug();

  const addBreakpoint = useDebugStore((state) => state.addBreakpoint);
  const removeBreakpoint = useDebugStore((state) => state.removeBreakpoint);
  const clearBreakpoints = useDebugStore((state) => state.clearBreakpoints);
  const breakpoints = useDebugStore((state) => state.breakpoints);

  const [breakpointNameAdd, setBreakpointNameAdd] = useState("");
  const [breakpointNameRemove, setBreakpointNameRemove] = useState("");
  const [breakpointListCollapsed, setBreakpointListCollapsed] = useState(true);

  const handleAddBreakpoint = () => {
    addBreakpoint(breakpointNameAdd);
    runDbgCmd(`break ${breakpointNameAdd}`, (data) => console.log({ data }));
  };

  const handleRemoveBreakpoint = () => {
    removeBreakpoint(breakpointNameRemove);
    runDbgCmd(`clear ${breakpointNameRemove}`, (data) => console.log({ data }));
  };

  const setRegisters = useDebugStore((state) => state.setRegisters);
  const setFlags = useDebugStore((state) => state.setFlags);

  const handleUpdateInfo = () => {
    runDbgCmd("info registers", (arg) => {
      const transformedResp = transformRegistersResponse(arg.data);

      console.log({ transformedResp });

      setRegisters(transformedResp.registers);
      setFlags(transformedResp.flags);

      console.log({ transformedResp });
    });
  };

  return (
    <div className="mt-2">
      <div className="mx-2 flex w-min items-center gap-2 rounded-md border bg-white p-1 font-bold">
        <Tooltip tooltip="Continue">
          <div
            onClick={() => runDbgCmd("run", () => null)}
            className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner"
          >
            <VscDebugStart />
          </div>
        </Tooltip>
        <Tooltip tooltip="Step Over">
          <div
            onClick={() => runDbgCmd("nexti", () => null)}
            className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner"
          >
            <VscDebugStepOver />
          </div>
        </Tooltip>
        <Tooltip tooltip="Step In">
          <div
            onClick={() => runDbgCmd("stepi", () => null)}
            className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner"
          >
            <VscDebugStepInto />
          </div>
        </Tooltip>
        <Tooltip tooltip="Update info">
          <div
            onClick={handleUpdateInfo}
            className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner"
          >
            <HiRefresh />
          </div>
        </Tooltip>
        <Tooltip tooltip="Stop" onClick={() => runDbgCmd("kill", () => null)}>
          <div className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner">
            <TbPlayerStop />
          </div>
        </Tooltip>
        <Tooltip tooltip="Exit" onClick={() => runDbgCmd("quit", () => null)}>
          <div className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner">
            <IoIosExit />
          </div>
        </Tooltip>
      </div>

      <div className="mt-2">
        <div className=" px-2 text-sm font-bold">Add breakpoint:</div>
        <div className="mx-2 flex items-stretch gap-2 rounded-md border p-1 font-bold">
          <Input
            value={breakpointNameAdd}
            onChange={(e) => setBreakpointNameAdd(e.target.value)}
            className="py-1 text-sm font-normal "
            placeholder="Line number or id..."
          />
          <Button
            onClick={handleAddBreakpoint}
            className="flex items-center px-2 py-1 text-sm active:outline-1"
          >
            Add
          </Button>
        </div>
        <div className="mx-2 flex items-stretch gap-2 rounded-md border p-1 font-bold">
          <Input
            value={breakpointNameRemove}
            onChange={(e) => setBreakpointNameRemove(e.target.value)}
            className="py-1 text-sm font-normal "
            placeholder="Line number or id..."
          />
          <Button
            onClick={handleRemoveBreakpoint}
            className="flex items-center px-2 py-1 text-sm active:outline-1"
          >
            Delete
          </Button>
        </div>
        <div className="mx-3 mt-1 flex">
          <Button
            onClick={clearBreakpoints}
            className="flex w-full items-center gap-2 px-2 py-1 text-sm active:outline-1"
          >
            <MdOutlineClear size={20} /> Clear all breakpoints
          </Button>
        </div>
        <div
          className="mx-3 mt-2  rounded bg-white p-1 text-sm font-semibold"
          onClick={() => setBreakpointListCollapsed((prev) => !prev)}
        >
          <div className="flex cursor-pointer items-center justify-between">
            Breakpoints list:
            <IoIosArrowForward
              className={clsxm(
                breakpointListCollapsed ? "rotate-90" : "-rotate-90",
                "transition-all",
              )}
            />
          </div>

          {!breakpointListCollapsed && (
            <div className="mt-1 text-xs font-normal">
              {breakpoints.map((breakpoint) => (
                <div>{breakpoint}</div>
              ))}
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

export default DebugControls;
