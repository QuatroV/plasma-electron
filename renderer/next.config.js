module.exports = {
  webpack: (config, { isServer }) => {
    if (!isServer) {
      config.target = 'electron-renderer';
    }
    
    config.module.rules.push({
      test: /\.svg$/i,
      issuer: /\.[jt]sx?$/,
      use: ['@svgr/webpack'],
    })

    return config;
  },
};
