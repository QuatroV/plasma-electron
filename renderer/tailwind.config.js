/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [
    './renderer/pages/**/*.{js,ts,jsx,tsx}',
    './renderer/components/**/*.{js,ts,jsx,tsx}',
    './renderer/features/**/*.{js,ts,jsx,tsx}',
  ],
  theme: {
    extend: {
      backgroundImage: {
        'gradient-radial': 'radial-gradient(var(--tw-gradient-stops))',
      },
      keyframes: {
        "slow-appear": {
          '0%': { transform: 'scale(0.1, 1)', transform: 'translate(-5%, 0)', opacity: "0.1" },
          '100%': { transform: 'scale(1, 1)', transform: 'translate(0, 0)', opacity: "1" },
        },
        "slow-appear-no-opacity": {
          '0%': { transform: 'scale(0.1, 1)', transform: 'translate(-5%, 0)'},
          '100%': { transform: 'scale(1, 1)', transform: 'translate(0, 0)'},
        },
        "roll": {
          from: { transform: 'rotate(0deg)'},
          to: { transform: 'rotate(-360deg)'},
        },
        "shake": {
          "10%":{
            transform: "translate3d(-1px, 0, 0)",
          },
          
          "20%":{
            transform: "translate3d(2px, 0, 0)",
          },
        
          "30%":{
            transform: "translate3d(-4px, 0, 0)",
          },
        
          "40%":{
            transform: "translate3d(4px, 0, 0)",
          },

          "50%":{
            transform: "translate3d(-4px, 0, 0)",
          },
        
          "60%":{
            transform: "translate3d(4px, 0, 0)",
          },

          "70%":{
            transform: "translate3d(4px, 0, 0)",
          },

          "80%":{
            transform: "translate3d(2px, 0, 0)",
          },

          "90%":{
            transform: "translate3d(-1px, 0, 0)",
          },
          
        },
        "scale": {
          from: { transform: 'scale(0)'},
          to: { transform: 'scale(1)'},
        }
      },
      animation: {
        'slow-appear': 'slow-appear 0.25s linear',
        "slow-appear-no-opacity": 'slow-appear-no-opacity 0.25s linear',
        'one-roll': 'roll .25s linear',
        'fast-shake': 'shake .25s linear',
        'scale': 'scale .12s linear',
      },
    },
    fontFamily: {
      rubik: ['Rubik', 'sans-serif'],
    }
  },
  plugins:[
    require('@tailwindcss/container-queries'),
  ],
};
