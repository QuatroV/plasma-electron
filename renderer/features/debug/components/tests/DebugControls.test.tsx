import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";

import DebugControls from "../DebugControls";

// Adjust this to the actual path of your component

describe("DebugControls", () => {
  // Hack, see link: https://github.com/facebook/react/issues/11565#issuecomment-380143358
  beforeAll(() => {
    //@ts-ignore
    ReactDOM.createPortal = jest.fn((element, node) => {
      return element;
    });
  });

  afterEach(() => {
    //@ts-ignore
    ReactDOM.createPortal.mockClear();
  });
  it("renders correctly", () => {
    const tree = renderer.create(<DebugControls />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
