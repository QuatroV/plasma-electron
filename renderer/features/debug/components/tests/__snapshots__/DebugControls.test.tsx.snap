// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DebugControls renders correctly 1`] = `
<div
  className="mt-2"
>
  <div
    className="mx-2 flex w-min items-center gap-2 rounded-md border bg-white p-1 font-bold"
  >
    <div
      className=""
    >
      <div
        className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M4.25 3l1.166-.624 8 5.333v1.248l-8 5.334-1.166-.624V3zm1.5 1.401v7.864l5.898-3.932L5.75 4.401z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      className="animate-scale rounded bg-gray-800 p-2 text-xs text-white shadow transition-all"
      tabIndex={-1}
    >
      Continue
    </div>
    <div
      className=""
    >
      <div
        className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M14.25 5.75v-4h-1.5v2.542c-1.145-1.359-2.911-2.209-4.84-2.209-3.177 0-5.92 2.307-6.16 5.398l-.02.269h1.501l.022-.226c.212-2.195 2.202-3.94 4.656-3.94 1.736 0 3.244.875 4.05 2.166h-2.83v1.5h4.163l.962-.975V5.75h-.004zM8 14a2 2 0 1 0 0-4 2 2 0 0 0 0 4z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      className="animate-scale rounded bg-gray-800 p-2 text-xs text-white shadow transition-all"
      tabIndex={-1}
    >
      Step Over
    </div>
    <div
      className=""
    >
      <div
        className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M8 9.532h.542l3.905-3.905-1.061-1.06-2.637 2.61V1H7.251v6.177l-2.637-2.61-1.061 1.06 3.905 3.905H8zm1.956 3.481a2 2 0 1 1-4 0 2 2 0 0 1 4 0z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      className="animate-scale rounded bg-gray-800 p-2 text-xs text-white shadow transition-all"
      tabIndex={-1}
    >
      Step In
    </div>
    <div
      className=""
    >
      <div
        className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M4 2a1 1 0 011 1v2.101a7.002 7.002 0 0111.601 2.566 1 1 0 11-1.885.666A5.002 5.002 0 005.999 7H9a1 1 0 010 2H4a1 1 0 01-1-1V3a1 1 0 011-1zm.008 9.057a1 1 0 011.276.61A5.002 5.002 0 0014.001 13H11a1 1 0 110-2h5a1 1 0 011 1v5a1 1 0 11-2 0v-2.101a7.002 7.002 0 01-11.601-2.566 1 1 0 01.61-1.276z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      className="animate-scale rounded bg-gray-800 p-2 text-xs text-white shadow transition-all"
      tabIndex={-1}
    >
      Update info
    </div>
    <div
      className=""
    >
      <div
        className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
            stroke="none"
          />
          <path
            d="M5 5m0 2a2 2 0 0 1 2 -2h10a2 2 0 0 1 2 2v10a2 2 0 0 1 -2 2h-10a2 2 0 0 1 -2 -2z"
          />
        </svg>
      </div>
    </div>
    <div
      className="animate-scale rounded bg-gray-800 p-2 text-xs text-white shadow transition-all"
      tabIndex={-1}
    >
      Stop
    </div>
    <div
      className=""
    >
      <div
        className="cursor-pointer rounded-md p-1 hover:bg-gray-100 active:shadow-inner"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M144 136v107h179.9l-47.2-47.9c-5-5.1-5-13.3.1-18.4 5.1-5 13.3-5 18.4.1l69 70c2.4 2.5 3.7 5.8 3.7 9.1 0 1.7-.3 3.4-1 5-.6 1.5-1.6 2.9-2.7 4.1l-69 70c-5 5.1-13.3 5.2-18.4.1-5.1-5-5.2-13.3-.1-18.4l47.2-47.9H144v107c0 22 18 40 40 40h240c22 0 40-18 40-40V136c0-22-18-40-40-40H184c-22 0-40 18-40 40zM61 243c-7.2 0-13 5.8-13 13s5.8 13 13 13h83v-26H61z"
          />
        </svg>
      </div>
    </div>
    <div
      className="animate-scale rounded bg-gray-800 p-2 text-xs text-white shadow transition-all"
      tabIndex={-1}
    >
      Exit
    </div>
  </div>
  <div
    className="mt-2"
  >
    <div
      className=" px-2 text-sm font-bold"
    >
      Add breakpoint:
    </div>
    <div
      className="mx-2 flex items-stretch gap-2 rounded-md border p-1 font-bold"
    >
      <input
        className="w-full rounded p-1 py-1 text-sm font-normal"
        onChange={[Function]}
        placeholder="Line number or id..."
        value=""
      />
      <button
        className="float-right rounded bg-white p-2 font-medium transition-all hover:shadow-lg active:outline active:outline-emerald-400 flex items-center px-2 py-1 text-sm active:outline-1"
        onAnimationEnd={[Function]}
        onClick={[Function]}
      >
        Add
      </button>
    </div>
    <div
      className="mx-2 flex items-stretch gap-2 rounded-md border p-1 font-bold"
    >
      <input
        className="w-full rounded p-1 py-1 text-sm font-normal"
        onChange={[Function]}
        placeholder="Line number or id..."
        value=""
      />
      <button
        className="float-right rounded bg-white p-2 font-medium transition-all hover:shadow-lg active:outline active:outline-emerald-400 flex items-center px-2 py-1 text-sm active:outline-1"
        onAnimationEnd={[Function]}
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
    <div
      className="mx-3 mt-1 flex"
    >
      <button
        className="float-right rounded bg-white p-2 font-medium transition-all hover:shadow-lg active:outline active:outline-emerald-400 flex w-full items-center gap-2 px-2 py-1 text-sm active:outline-1"
        onAnimationEnd={[Function]}
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height={20}
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
          />
        </svg>
         Clear all breakpoints
      </button>
    </div>
    <div
      className="mx-3 mt-2  rounded bg-white p-1 text-sm font-semibold"
      onClick={[Function]}
    >
      <div
        className="flex cursor-pointer items-center justify-between"
      >
        Breakpoints list:
        <svg
          className="rotate-90 transition-all"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M294.1 256L167 129c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.3 34 0L345 239c9.1 9.1 9.3 23.7.7 33.1L201.1 417c-4.7 4.7-10.9 7-17 7s-12.3-2.3-17-7c-9.4-9.4-9.4-24.6 0-33.9l127-127.1z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
